client/order/orderbook.go

```diff
@@ -0,0 +1,226 @@
+package order
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+
+	"github.com/decred/dcrdex/dex/msgjson"
+)
+
+// RemoteOrderBook defines the functions a client tracked order book
+// must implement.
+type RemoteOrderBook interface {
+	// Initialize instantiates a client tracked order book with the
+	// current order book snapshot.
+	Initialize(*msgjson.OrderBook)
+	// Book adds a new order to the order book.
+	Book(*msgjson.BookOrderNote)
+	// Unbook removes an order from the order book.
+	Unbook(*msgjson.UnbookOrderNote) error
+}
+
+// OrderBook represents an client tracked order book.
+type OrderBook struct {
+	seq      uint64
+	marketID string
+	orders   map[string]*msgjson.BookOrderNote
+	buys     map[uint64][]*msgjson.BookOrderNote
+	sells    map[uint64][]*msgjson.BookOrderNote
+	synced   bool
+	mtx      sync.RWMutex
+}
+
+// addOrder adds an order to the order book. This must be called with
+// the order book mutex held.
+func (ob *OrderBook) addOrder(order *msgjson.BookOrderNote) {
+	ob.orders[order.OrderID.String()] = order
```

Comment:
We could use another type like `msgjson.Bytes` but for a `[32]byte`. Maybe `type Hash256 [32]byte`?

Reply:
Maybe.  I just don't recall the reason for a slice instead of an array, if there was one.  Having it static length where it works is all around better.

Reply:
I'll plan on adding a `[32]byte` `msgjson` type. I hesitate to try to satisfy `json.Marshaler` and `json.Unmarshaler` directly with the `order.OrderID` and `order.MatchID` and import those into `msgjson`. Though I believe that is possible without any dependency issues, is creates a hierarchy in those packages that may not be desirable at a later stage. 

@dnldd: in the meantime, please use `order.OrderID` as the key here, because `string` keys are both longer and a lot slower in Go (though in practice, probably not a noticeable performance difference on the client). You'll need to add an error for now, but we can remove it once the new `msgjson` type is implemented. Use something like 

```
if len(order.OrderID) != 32 {
    return fmt.Errorf(...)
}
var oid *order.OrderID
copy(oid[:], order.OrderID)
```

to get the order ID.

Reply:
sure thing, thanks